File: Shloka and the Broken Lift/gemini1.cpp
#include <iostream>
#include <vector>

using namespace std;

void solve() {
    int n;
    cin >> n;
    vector<int> h(n + 1);
    for (int i = 1; i <= n; i++) {
        cin >> h[i];
    }

    vector<bool> dp(n + 1, false);
    dp[1] = true;

    for (int j = 2; j <= n; j++) {
        // Find factors of h[j] up to sqrt(h[j])
        for (int d = 1; d * d <= h[j]; d++) {
            if (h[j] % d == 0) {
                // Check jump distance d
                if (d >= 2 && j - d >= 1 && dp[j - d]) {
                    dp[j] = true;
                    break;
                }
                // Check the paired factor (h[j] / d)
                int paired_d = h[j] / d;
                if (paired_d >= 2 && j - paired_d >= 1 && dp[j - paired_d]) {
                    dp[j] = true;
                    break;
                }
            }
        }
    }

    if (dp[n]) cout << "YES\n";
    else cout << "NO\n";
}

int main() {
    // Fast I/O
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int t;
    if (cin >> t) {
        while (t--) {
            solve();
        }
    }
    return 0;
}

File: Shloka and the Broken Lift/gemini1.py
import sys
import math

def solve():
    # Read all tokens from standard input at once
    input_data = sys.stdin.read().split()
    if not input_data:
        return
    
    t = int(input_data[0])
    idx = 1
    
    out = []
    
    for _ in range(t):
        n = int(input_data[idx])
        idx += 1
        
        # 1-based indexing for convenience 
        h = [0] + [int(x) for x in input_data[idx : idx+n]]
        idx += n
        
        dp = [False] * (n + 1)
        dp[1] = True
        
        for j in range(2, n + 1):
            limit = math.isqrt(h[j])
            for d in range(1, limit + 1):
                if h[j] % d == 0:
                    # Check factor 'd'
                    if d >= 2 and j - d >= 1 and dp[j - d]:
                        dp[j] = True
                        break
                    
                    # Check paired factor
                    paired_d = h[j] // d
                    if paired_d >= 2 and j - paired_d >= 1 and dp[j - paired_d]:
                        dp[j] = True
                        break
                        
        if dp[n]:
            out.append("YES")
        else:
            out.append("NO")
            
    print('\n'.join(out))

if __name__ == '__main__':
    solve()